
**Comentario del Revisor**

Hola!

Soy Juan Manuel Romero, pero siéntete libre de llamarme Juanma. Soy code reviewer en Tripleten y hoy estaré revisando tu entrega.

Para simular la dinámica de un ambiente de trabajo, si veo algún error, en primer instancia solo los señalaré, dándote la oportunidad de encontrarlos y corregirlos por tu cuenta. En un trabajo real, el líder de tu equipo hará una dinámica similar. En caso de que no puedas resolver la tarea, te daré una información más precisa en la próxima revisión. 

Solo un aviso rápido: cuando estés revisando el proyecto, por favor deja mis comentarios originales tal como están. De esta manera, podemos seguir fácilmente el progreso y asegurarnos de que no se nos pase nada por alto. Y, si realizas algún cambio basado en mis comentarios, sería genial si pudieras resaltar esas actualizaciones para que se destaquen.

Puedes encontrar mis comentarios en cajas verdes, amarillas o rojas como estas:

<div class="alert alert-block alert-success"> 
<b>Comentario del Revisor</b> <a class="tocSkip"></a>

Éxito. Todo se ha hecho correctamente.

</div>


<div class="alert alert-block alert-warning"> 
<b>Comentario del Revisor</b> <a class="tocSkip"></a>

Observaciones. Algunas recomendaciones.

</div> 


<div class="alert alert-block alert-danger">
<b>Comentario del Revisor</b> <a class="tocSkip"></a>

Requiere corrección. El bloque requiere algunas correcciones. El trabajo no puede ser aceptado con los comentarios en rojo.

</div>

Puedes responderme usando esto:

<div class="alert alert-block alert-info"> <b>Respuesta del estudiante.</b> <a class="tocSkip"></a> </div>

## comenzamos por importar las librerias que seran utilizadas

In [14]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score
import numpy as np

## leemos los datos despues de descargarlos

In [15]:
data =  pd.read_csv('/datasets/users_behavior.csv')


print(data.head())
print(data.info())
print(data.describe())

   calls  minutes  messages   mb_used  is_ultra
0   40.0   311.90      83.0  19915.42         0
1   85.0   516.75      56.0  22696.96         0
2   77.0   467.66      86.0  21060.45         0
3  106.0   745.53      81.0   8437.39         1
4   66.0   418.74       1.0  14502.75         0
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3214 entries, 0 to 3213
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   calls     3214 non-null   float64
 1   minutes   3214 non-null   float64
 2   messages  3214 non-null   float64
 3   mb_used   3214 non-null   float64
 4   is_ultra  3214 non-null   int64  
dtypes: float64(4), int64(1)
memory usage: 125.7 KB
None
             calls      minutes     messages       mb_used     is_ultra
count  3214.000000  3214.000000  3214.000000   3214.000000  3214.000000
mean     63.038892   438.208787    38.281269  17207.673836     0.306472
std      33.236368   234.569872    36.148326   7570.968246  

<div class="alert alert-block alert-success"> 
<b>Comentario del Revisor #1</b> <a class="tocSkip"></a>

Buen trabajo! Has utilizado métodos como _info()_, _head()_ o _describe()_ para analizar los datos. Estos métodos te permiten entender la estructura de los datos, los tipos de variables, la presencia de valores faltantes y obtener una descripción estadística general.

</div>

aqui podemos observar que los datos ya han sido procesados y estan listos para utilizarse en un modelo

In [16]:
#asignamos features y target
features = data.drop(['is_ultra'], axis=1)
target = data['is_ultra']

## ahora segmentamos los datos en target, train y test

<div class="alert alert-block alert-success"> 
<b>Comentario del Revisor #1</b> <a class="tocSkip"></a>

Excelente! Has dividido correctamente los datos en tres conjuntos: _train_, _validation_ y _test_. Dividir los datos de esta forma permite entrenar el modelo en un conjunto, validar su desempeño en otro y finalmente evaluarlo en un conjunto independiente para obtener una medida precisa de su rendimiento.
    
Has utilizado correctamente _train_test_split_ dos veces para dividir los datos en tres conjuntos (entrenamiento, validación y prueba). Este método es el más común para dividir los datos de manera aleatoria, asegurando que cada conjunto tenga una representación justa.

</div>

corregimos test_size en la siguiente celda

In [17]:


features_train, features_valid, target_train, target_valid = train_test_split(features, target, test_size=0.4, random_state=54321)
features_valid, features_test, target_valid, target_test = train_test_split(features_valid, target_valid, test_size=0.5, random_state=54321)

print(f"Tamaño de features_train: {features_train.shape}")
print(f"Tamaño de target_train: {target_train.shape}")
print(f"Tamaño de features_valid: {features_valid.shape}")
print(f"Tamaño de target_valid: {target_valid.shape}")
print(f"Tamaño de features_test: {features_test.shape}")
print(f"Tamaño de target_test: {target_test.shape}")


Tamaño de features_train: (1928, 4)
Tamaño de target_train: (1928,)
Tamaño de features_valid: (643, 4)
Tamaño de target_valid: (643,)
Tamaño de features_test: (643, 4)
Tamaño de target_test: (643,)


<div class="alert alert-block alert-danger"> 
<b>Comentario del Revisor #1</b> <a class="tocSkip"></a>

Parece que el conjunto de test es demasiado pequeño. Para una correcta validación de los modelos, debe ser al menos el 10%, idealmente el 20%, para garantizar que se pueda hacer una buena evaluación de los hiperparámetros. Asegúrate de reservar una cantidad adecuada de datos para esta etapa.
    
TIP: Puedes aumentar el porcentaje de test_size en el segundo train_test_split.
    
Te recomiendo que imprimas el _shape_ de cada conjunto para que te des cuenta de sus tamaños finales. Quizás un buen corte para este proyecto sería 60 - 20 - 20

</div>

<div class="alert alert-block alert-success"> 
<b>Comentario del Revisor #2</b> <a class="tocSkip"></a>

Ahora sí! Los tamaños de los conjuntos son correctos. Excelente trabajo.

</div>

## probamos regresion lineal

agregamos regresion lineal

In [18]:
# Entrenamiento del modelo de Regresión Logística
logistic_model = LogisticRegression(random_state=54321, max_iter=1000)
logistic_model.fit(features_train, target_train)

# Evaluación en conjunto de validación
logistic_valid_predictions = logistic_model.predict(features_valid)
logistic_accuracy = accuracy_score(target_valid, logistic_valid_predictions)

print(f'Precisión de Regresión Logística (Validación): {logistic_accuracy:.4f}')

Precisión de Regresión Logística (Validación): 0.6750


## probamos con arbol de decision

agregamos arbol de desicion

In [19]:
tree_model = DecisionTreeClassifier(random_state=54321, max_depth=10)
tree_model.fit(features_train, target_train)

# Evaluación en conjunto de validación
tree_valid_predictions = tree_model.predict(features_valid)
tree_accuracy = accuracy_score(target_valid, tree_valid_predictions)

print(f'Precisión del Árbol de Decisión (Validación): {tree_accuracy:.4f}')

Precisión del Árbol de Decisión (Validación): 0.7823


<div class="alert alert-block alert-success"> 
<b>Comentario del Revisor #2</b> <a class="tocSkip"></a>

Genial! Consideraste varios modelos para el ajuste. 
</div>

DecisionTreeClassifier nos da un buen porcentaje con un 78.23%

## probamos RandomForestClassifier 
porque es un modelo muy sólido que ofrece buenos resultados para clasificación sin necesidad de un ajuste complejo, y es menos propenso a sobreajustarse y a ser afectado por datos ruidosos o correlacionados.

<div class="alert alert-block alert-success"> 
<b>Comentario del Revisor #1</b> <a class="tocSkip"></a>

Has entrenado el modelo únicamente con el conjunto de entrenamiento, lo que es exactamente lo que se debe hacer. Este conjunto contiene los datos con los que el modelo aprende los patrones, y no debe utilizar otros conjuntos durante el entrenamiento.

Luego, has validado correctamente el modelo utilizando solo el conjunto de validación. Esta es una práctica esencial para ajustar los hiperparámetros y comparar diferentes modelos sin introducir sesgos.


</div>

In [20]:

model = RandomForestClassifier(random_state=54321)

# Entrenar el modelo
model.fit(features_train, target_train)

# Predecir en el conjunto de validación
target_valid_pred = model.predict(features_valid)

# Evaluar la precisión en el conjunto de validación
accuracy_valid = accuracy_score(target_valid, target_valid_pred)
print(f'Precisión en la validación: {accuracy_valid:.4f}')


Precisión en la validación: 0.7823


la precisión en la validación es 0.7823, lo que significa que el 78.23% de las predicciones realizadas por el modelo en el conjunto de validación fueron correctas. Nos dio lo mismo que DecisionTreeClassifier y decidimos quedarnos con RandomForestClassifier

<div class="alert alert-block alert-info"> <b>Utilizamos un bucle for para iterar e posibles valores de hiperparametros y escoger el mejor modelo.</b> <a class="tocSkip"></a> </div>

In [29]:
best_score_tree = 0
best_params_tree = {}

for depth in range(1, 21):
    for criterion in ['gini', 'entropy']:
        tree_model = DecisionTreeClassifier(max_depth=depth, criterion=criterion, random_state=54321)
        tree_model.fit(features_train, target_train)
        predictions_valid = tree_model.predict(features_valid)
        score = accuracy_score(target_valid, predictions_valid)
        
        if score > best_score_tree:
            best_score_tree = score
            best_params_tree = {'max_depth': depth, 'criterion': criterion}

print(f"Mejores parámetros para DecisionTreeClassifier: {best_params_tree}")
print(f"Mejor puntuación en validación: {best_score_tree:.4f}")

# Ajuste de hiperparámetros para RandomForestClassifier
best_score_forest = 0
best_params_forest = {}

for n_estimators in range(10, 201, 10):  # Probar de 10 a 100 árboles
    for depth in range(5, 21, 5):        # Probar profundidades de 5 a 20
        model = RandomForestClassifier(n_estimators=n_estimators, max_depth=depth, random_state=54321)
        model.fit(features_train, target_train)
        predictions_valid = model.predict(features_valid)
        score = accuracy_score(target_valid, predictions_valid)
        
        if score > best_score_forest:
            best_score_forest = score
            best_params_forest = {'n_estimators': n_estimators, 'max_depth': depth}

print(f"Mejores parámetros para RandomForestClassifier: {best_params_forest}")
print(f"Mejor puntuación en validación: {best_score_forest:.4f}")

Mejores parámetros para DecisionTreeClassifier: {'max_depth': 10, 'criterion': 'gini'}
Mejor puntuación en validación: 0.7823
Mejores parámetros para RandomForestClassifier: {'n_estimators': 10, 'max_depth': 10}
Mejor puntuación en validación: 0.8025


<div class="alert alert-block alert-success"> 
<b>Comentario del Revisor #3</b> <a class="tocSkip"></a>

Impecable!! Gran trabajo Nicolas.

</div>


podemos observar que el mejor modelo es el de RandomForestClassifier y los mejores parametros que podemos usar son n_estimators:10, max_depth:10 para conseguir un 80.25% de aciertos

## nos quedamos con RandomForestClassifier

In [31]:
model = RandomForestClassifier(n_estimators=10, max_depth=10, random_state=54321)

# Reentrenar el modelo
model.fit(features_train, target_train)

# Predecir en el conjunto de validación
target_valid_pred = model.predict(features_valid)

# Evaluar la precisión en el conjunto de validación
accuracy_valid = accuracy_score(target_valid, target_valid_pred)
print(f'Precisión ajustada en la validación: {accuracy_valid:.4f}')


Precisión ajustada en la validación: 0.8025


El modelo ha sido ajustado y ha mejorado ligeramente su precisión en la validación, alcanzando un 80.25% de aciertos

In [32]:

target_test_pred = model.predict(features_test)

# Evaluar la precisión en el conjunto de prueba
accuracy_test = accuracy_score(target_test, target_test_pred)
print(f'Precisión en el conjunto de prueba: {accuracy_test:.4f}')


Precisión en el conjunto de prueba: 0.8243


El modelo tiene una precisión del 82.43% en los datos de prueba, lo que indica que el modelo tiene un rendimiento bastante bueno incluso en un conjunto de datos no visto previamente.

In [33]:
if accuracy_test >= 0.75:
    print("El modelo ha alcanzado la precisión mínima requerida")
else:
    print("El modelo no ha alcanzado la precisión mínima.")

El modelo ha alcanzado la precisión mínima requerida


<div class="alert alert-block alert-success"> 
<b>Comentario del Revisor #1</b> <a class="tocSkip"></a>

Excelente! Has logrado una precisión (_accuracy_) superior a 0.75, tal como se pide.

</div>

Se evalúa la calidad del modelo mediante la precisión en el conjunto de validación y prueba. El código también incluye una verificación de la precisión mínima de 0.75 en el conjunto de prueba

<div class="alert alert-block alert-success"> 
<b>Comentario del Revisor #1</b> <a class="tocSkip"></a>

Has documentado los resultados de manera clara, lo que es crucial para comunicar los hallazgos y justificar la elección del mejor modelo. Explicar el razonamiento detrás de tus decisiones es clave para asegurar que los resultados sean interpretados correctamente.

</div>

<div class="alert alert-block alert-danger"> 
<b>Comentario del Revisor #1</b> <a class="tocSkip"></a>

Parece que solo consideraste un modelo. Es importante probar más de uno para comparar su rendimiento. Al probar diferentes algoritmos, puedes descubrir qué modelo se ajusta mejor a la naturaleza de los datos y los objetivos del proyecto.
    
Además, no ajustaste suficientes hiperparámetros. Es fundamental ajustar al menos tres hiperparámetros para optimizar el rendimiento de cada modelo. Intenta explorar más opciones de ajuste. Puedes utilizar bucles _for_ para iterar en algunos posibles valores y seleccionar el mejor conjunto de hiperparámetros validando con el conjunto de validación.


</div>

## Conclusión general del proyecto
En este proyecto, desarrolle un modelo de clasificación usando RandomForestClassifier para predecir el plan adecuado  para los usuarios de Megaline. El modelo alcanzó una precisión del 82.89% en el conjunto de prueba, superando el umbral mínimo de 0.75.

El proceso incluyó segmentación en tres conjuntos, y la optimización de hiperparámetros. El modelo mostró buenos resultados, cumpliendo con los requisitos y demostrando que RandomForestClassifier es una herramienta efectiva para esta tarea. Estoy muy satisfecho con los resultados obtenidos.


<div class="alert alert-block alert-success"> 
<b>Comentario del Revisor #1</b> <a class="tocSkip"></a>

Muy buen trabajo Nicolas! 

<div class="alert alert-block alert-danger">
<b>Comentario del Revisor #1</b> <a class="tocSkip"></a>

Te he dejado algunos comentarios, pero para poder darlo por aprobado debes completar los pasos de la siguiente lista que no estan marcados como completos:

</div>

- [ ]  El tamaño del conjunto de entrenamiento es mayor al 50% del conjunto completo (Idealmente el 60%).
- [ ]  El tamaño del conjunto de validación es mayor al 10% del conjunto completo (Idealmente el 20%).
- [ ]  El tamaño del conjunto de test es mayor al 10% del conjunto completo (Idealmente el 20%)
- [ ]  Se ha realizado el ajuste del modelo.
    - [ ]  Se ha considerado más de un modelo.
    - [ ]  Se han considerado al menos 3 hiperparámetros para cada modelo.
- [ ]  Una vez elegido los hiperparámetros y elegido el mejor modelo, se testea el modelo con el conjunto de test (Sólo debe utilizarse en esta etapa).

</div>

<div class="alert alert-block alert-info"> <b>Cambie los tamaños de test_size e imprimi shape, tambien probe los model de arbol de decision, regresion linear y bosque aleatorio para poder finalmente escoger la mejor opcion. Espero haber corregido todo lo necesario.</b> <a class="tocSkip"></a> </div>

<div class="alert alert-block alert-success"> 
<b>Comentario del Revisor #2</b> <a class="tocSkip"></a>

Muy buen avance Nicolas! 

<div class="alert alert-block alert-danger"> 
<b>Comentario del Revisor #2</b> <a class="tocSkip"></a>

Solo te falta realizar el ajuste de hiperparámetros de los modelos, al menos para RandomForestClassifier y DecisionTreeClassifier. Esto te permite ajustar aún más los modelos y lograr, posiblemente, mejores resultados. 
    
Este paso es muy importante en esta proyecto. 
    
Puedes utilizar un bucle _for_ para iterar en posibles valores de los hiperparámetros y elegir los mejores.

</div>

</div>

<div class="alert alert-block alert-info"> <b>Ahora si agregue un bucle para iterar en los parametros y pude observar que definitivamente esa es una mejor opcion ya que pude encontrar los mejores parametros para el mejor modelo. Ahora si espero haber corregido todo lo necesario y gracias por la atencion al revisar.</b> <a class="tocSkip"></a> </div>


<div class="alert alert-block alert-success"> 
<b>Comentario del Revisor #3</b> <a class="tocSkip"></a>

Felicidades Nicolas! Ahora si, tu trabajo es excelente. 

Has completado todos los items necesarios para aprobar la entrega.

</div>